### PR TITLE
Include note of common profiling issue

### PR DESCRIPTION
**Pull request by tbillington**
- Includes a note about a common issue seen when profiling with span based tracing tools
- Adds a table of contents to profile.md

[Rendered](https://github.com/tbillington/bevy/blob/profiling-doc-update/docs/profiling.md).

[Suggested](https://discord.com/channels/691052431525675048/866787577687310356/1141612079460139108) by Jasmine on discord.

### CODE REVIEW

**Review by alice-i-cecile — APPROVED** [review]
This is a nice improvement! Both the common issue and the cleanup are nice.

File: docs/profiling.md [hunk 1]
```diff
@@ -1,22 +1,39 @@
...
+- [Compile time](#compile-time)
+
+# Runtime
 
 Bevy has built-in [tracing](https://github.com/tokio-rs/tracing) spans to make it cheap and easy to profile Bevy ECS systems, render logic, engine internals, and user app code. Enable the `trace` cargo feature to enable Bevy's built-in spans.
 
 If you also want to include `wgpu` tracing spans when profiling, they are emitted at the `tracing` `info` level so you will need to make sure they are not filtered out by the `LogSettings` resource's `filter` member which defaults to `wgpu=error`. You can do this by setting the `RUST_LOG=info` environment variable when running your application.
 
-You also need to select a `tracing` backend using the following cargo features:
+You also need to select a `tracing` backend using the following cargo features.
```
Comment (JMS55) [inline docs/profiling.md:18]:
```suggestion
You also need to select a `tracing` backend using one of the following cargo features.
```

File: docs/profiling.md [hunk 2]
```diff
@@ -1,22 +1,39 @@
...
+
+**⚠️ Note: for users of span based profilers**
+
+When GPU bottlenecked you may encounter frames that have multiple prepare-set systems all taking an unusually long time to complete, and all finishing at about the same time.
```
Comment (JMS55) [inline docs/profiling.md:22]:
```suggestion
When your app is bottlenecked by the GPU, you may encounter frames that have multiple prepare-set systems all taking an unusually long time to complete, and all finishing at about the same time.
```

**Review by JMS55 — CHANGES_REQUESTED** [review]
Thanks for writing this! Some minor wording changes and then we're good to go.

**Review by JMS55 — APPROVED** [review]

File: docs/profiling.md [hunk 3]
```diff
@@ -1,22 +1,39 @@
...
+## Table of Contents
+
+- [Runtime](#runtime)
+  - [Chrome tracing format](#chrome-tracing-format)
+  - [Tracy profiler](#tracy-profiler)
+  - [Adding your own spans](#adding-your-own-spans)
+  - [Perf flame graph](#perf-flame-graph)
+- [Compile time](#compile-time)
+
+# Runtime
```
Comment (mockersf) [inline docs/profiling.md:12]:
```suggestion
## Runtime
```

File: docs/profiling.md [hunk 4]
```diff
@@ -1,22 +1,39 @@
...
+**⚠️ Note: for users of span based profilers**
+
+When your app is bottlenecked by the GPU, you may encounter frames that have multiple prepare-set systems all taking an unusually long time to complete, and all finishing at about the same time.
+
+Improvements are planned to resolve this issue, you can find more details in the doc comment for [`prepare_systems`](../crates/bevy_render/src/view/window/mod.rs).
+
+![prepare_windows span bug](https://github.com/bevyengine/bevy/assets/2771466/15c0819b-0e07-4665-aa1e-579caa24fece)
 
-### Backend: trace_chrome
+## Chrome tracing format
```
Comment (mockersf) [inline docs/profiling.md:28]:
```suggestion
### Chrome tracing format
```

File: docs/profiling.md [hunk 5]
```diff
@@ -1,22 +1,39 @@
...
 
 `cargo run --release --features bevy/trace_chrome`
 
 After running your app a `json` file in the "chrome tracing format" will be produced. You can open this file in your browser using <https://ui.perfetto.dev>. It will look something like this:
 
 ![image](https://user-images.githubusercontent.com/2694663/141657409-6f4a3ad3-59b6-4378-95ba-66c0dafecd8e.png)
 
-### Backend: trace_tracy
+## Tracy profiler
```
Comment (mockersf) [inline docs/profiling.md:36]:
```suggestion
### Tracy profiler
```

File: docs/profiling.md [hunk 6]
```diff
@@ -52,7 +69,7 @@ If you save more than one trace, you can compare the spans between both of them
 
 ![A graph and statistics in the Tracy GUI comparing the distribution of execution times of an instrumented span across two traces](https://user-images.githubusercontent.com/3137680/205834698-84405b2f-97b5-43a3-9dba-385167ac1db5.png)
 
-### Adding your own spans
+## Adding your own spans
```
Comment (mockersf) [inline docs/profiling.md:72]:
```suggestion
### Adding your own spans
```

File: docs/profiling.md [hunk 7]
```diff
@@ -78,7 +95,7 @@ Search for `info_span!` in this repo for some real-world examples.
 
 For more details, check out the [tracing span docs](https://docs.rs/tracing/*/tracing/span/index.html).
 
-## `perf` Runtime Flame Graph
+## `perf` Flame Graph
```
Comment (mockersf) [inline docs/profiling.md:98]:
```suggestion
### `perf` Flame Graph
```

File: docs/profiling.md [hunk 8]
```diff
@@ -90,7 +107,7 @@ Install [cargo-flamegraph](https://github.com/flamegraph-rs/flamegraph), [enable
 After closing your app, an interactive `svg` file will be produced:
 ![image](https://user-images.githubusercontent.com/2694663/141657609-0089675d-fb6a-4dc4-9a59-871e95e31c8a.png)
 
-## Project Compile Times
+# Compile time
```
Comment (mockersf) [inline docs/profiling.md:110]:
```suggestion
## Compile time
```

**Review by mockersf — CHANGES_REQUESTED** [review]
You need to increase all your heading levels by 1, there can be only one `#`-level heading.

The linter is also unhappy about your note: https://github.com/updownpress/markdown-lint/blob/master/rules/036-no-emphasis-as-header.md

**Comment by mockersf** [comment]
it seems it's the first time the linter passed on this file, and it's unhappy with the list made with `*`, could you replace it with `-`?

**Comment by tbillington** [comment]
I was following the patterns in https://raw.githubusercontent.com/bevyengine/bevy/v0.11.2/examples/README.md.

Will update them in the morning, didn't know about the markdown linter, thanks!

File: docs/profiling.md [hunk 9]
```diff
@@ -1,22 +1,39 @@
...
 If you also want to include `wgpu` tracing spans when profiling, they are emitted at the `tracing` `info` level so you will need to make sure they are not filtered out by the `LogSettings` resource's `filter` member which defaults to `wgpu=error`. You can do this by setting the `RUST_LOG=info` environment variable when running your application.
 
-You also need to select a `tracing` backend using the following cargo features:
+You also need to select a `tracing` backend using one of the following cargo features.
+
+**⚠️ Note: for users of span based profilers**
+
+When your app is bottlenecked by the GPU, you may encounter frames that have multiple prepare-set systems all taking an unusually long time to complete, and all finishing at about the same time.
+
+Improvements are planned to resolve this issue, you can find more details in the doc comment for [`prepare_systems`](../crates/bevy_render/src/view/window/mod.rs).
```
Comment (Selene-Amanita) [inline docs/profiling.md:24]:
```suggestion
Improvements are planned to resolve this issue, you can find more details in the doc comment for [`prepare_windows`](../crates/bevy_render/src/view/window/mod.rs).
```

Also, I'm not sure I understand what is a "prepare-set" system, or what a swapchain is, what prepare_windows does, etc... Also I had to go down to "Adding your own spans" to understand what a span is, through context, so:
- maybe this doc should explain things a bit more?
- I don't think I'm qualified to review this PR ^^'

Comment (tbillington) [reply docs/profiling.md]: Nice catch, included it in the fixes!

I added a link to the definition of span in `tracing`.

As for the rest, it's a bit out of scope tbh, I just wanted to add a warning to this doc, I don't know what all of those are myself yet :)

**Review by mockersf — APPROVED** [review]

**Pull request merged**